### DISCUSSION TITLE
How to remove "url" & "input" fields from validation error response?

### DISCUSSION THREAD

**Discussion by KStenK** [pydantic/pydantic#7485] [Question] [answered]
Hi!

I'm using FastApi and am currently migrating our backend services slowly to Pydantic v2.

I discovered that when I get a validation error, the response shows unnecessary information, which I want to hide.

For example, the error response looks like this:

```
{'detail': [{
  'type': 'missing',
  'loc': ['query', 'session'],
  'msg': 'Field required',
  'input': None,
  'url': 'https://errors.pydantic.dev/2.3/v/missing'
}]}
```

I do not want to show 'url' and 'input' fields.

How can I disable or hide those fields?

**Comment by samuelcolvin** [comment, marked as answer]
Come on - how much research did you do before creating this question?

`url` can be omitted with [`errors(include_url=False)`](https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core._pydantic_core.ValidationError.errors).

There's a discussion about optionally omitting `input` at #7461.

Next time please RTFM.

**Reply by KStenK** [reply]
Okei, I understand that I can turn off the 'input' field with the **hide_input_in_errors** using model_config.

I understand how to disable the 'url' field when I'm using vanilla Pydantic, but I don't understand how to pass the **include_url** argument to the ValidationError.json() method when Fastapi is doing the error-handling magic behind the scenes.

For example, if the route looks something like this:

```
from fastapi import FastAPI
from pydantic import UUID4

app = FastAPI()


@app.get("/something")
def something(session: UUID4):
    . . .

```

Or is this question more suitable in the FastAPI repository?

**Reply by luckydonald** [reply]
Additionally, how do I do that in stack traces, without needing to build my own exception from scratch after having a bunch of json errors?

@samuelcolvin Also, your link doesn't lead anywhere remotely helpful. Gotta love pydantic.

**Reply by StuartBertram** [reply]
Also, RTFM is fine until it comes to doing this universally. Is there really not a generalised way to remove the input string from _all_ stack traces? It wasn't in the v1 stack traces when you stringified the exception.

(The correct doc link is https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core.ValidationError.errors and it's just for cases where you have a `ValidationError` and want to manually and explicitly call the `errors()` function to get a collection of errors)

**Comment by a-was** [comment]
Here's the discussion in the FastAPI repo
https://github.com/tiangolo/fastapi/discussions/10352

**Comment by luckydonald** [comment]
Here's a universal answer, which is only hinted about in the source code and not the "great" documentation:

```py
# PUT THIS BEFORE ANYTHING WHICH COULD DO A ValidationError !
import os
os.environ['PYDANTIC_ERRORS_INCLUDE_URL'] = '0'
```

Pydantic apparently don't wants people to _actually_ disable that.
Because to would be way too easy to just allow something like `pydantic.include_urls = False`... 
But that makes sense: It's nasty **user-facing advertisement for pydantic**, it spams the pydantic domain multiple times, one for each error. They seem to want to keep it that way.

**Reply by Kilo59** [reply]
You could create a simple context manager to prevent accidentally leaving this set (or unset).

```python
import contextlib
from typing import Generator
import os

@contextlib.contextmanager
def pydantic_errors_exclue_url() -> Generator[None, None, None]:
    """
    Temporarily exclude URL from error messages using the `PYDANTIC_ERRORS_INCLUDE_URL` environment variable.
    Reset the environment variable to its original value after the context manager exits.
    """
    missing = object() # Unique value to indicate that the environment variable was not set
    initial = os.environ.get('PYDANTIC_ERRORS_INCLUDE_URL', missing)
    os.environ['PYDANTIC_ERRORS_INCLUDE_URL'] = '0'
    yield
    if initial is missing:
        os.environ.pop('PYDANTIC_ERRORS_INCLUDE_URL', None)
    else:
        os.environ['PYDANTIC_ERRORS_INCLUDE_URL'] = initial

...

with pydantic_errors_exclude_url():
   something_that_raises_a_pydantic_error()
# back to default pydantic error behavior   
```

**Reply by luckydonald** [reply]
@Kilo59 Thanks, that would be very helpful, _if_ it would work.
Unfortunately, to quote the docs on `__repr__()`:
> Due to implementation details, this environment variable can only be set **once, before** the first validation error is created.

https://github.com/pydantic/pydantic-core/blob/562fad3e34805c3d5a2c2e8aec8d8d0193695a8e/python/pydantic_core/_pydantic_core.pyi#L667-L677

So yours might not even work if there was any other validation error before.

**Reply by luckydonald** [reply]
The only reason this option exists _at all_ is _not_ to give the user the choice to turn of the url spam,
but (_get this_) because the version string in the url would make it harder to write unittests...

To quote:
https://github.com/pydantic/pydantic/blob/aea47de182bbc0797446deaa91f7af7ddcfb0abc/tests/conftest.py#L58-L60

**Reply by ddmunhoz** [reply]
I can confirm. The env var is still the best way to get this done. Just had the same issue and wasted time checking the official docs... Thanks Donald!
